Exact Minimum Vertex Cover Algorithm

nolberg opened the pull request:
Added a method that computes an optimal vertex cover of an undirected graph.

jkinable commented:
Thank you for your submission. Jgrapht could indeed benefit from an exact algorithm for Vertex Covers. When looking into your implementation, I do have some concerns.
1. You recurse on subgraphs. Creating a subgraph is not exactly cheap (it takes O(V+E) time to create a subgraph which is a copy of your original graph. You may want to use more efficient data structures to keep track of which vertices are in your cover. There's no need to copy the entire graph structure. After all, you only need 1 data structure to track the neighbors of a vertex, and another to keep track of the vertices in your cover.
2. The recursive algorithm has the Overlapping Subproblems property. You want to apply memoization to make your implementation much more efficient.
3. Instead of traversing through the vertices at random, would it make sense to sort them?

I haven't looked into exact Vertex Cover implementations myself, but since this is such a well-studied problem I wouldn't be surprised if there are much more efficient implementations. Although you are obviously not expected to implement them, it would be great if you could mention that this is a rather simple implementation and that there exist much more efficient algorithms, namely....

Finally, please add unit tests in VertexCoversTest.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/VertexCovers.java:
```diff
@@ -147,6 +147,60 @@
     }
+    
+    /**
+     * Finds an optimal vertex cover of a specified graph. At each iteration,
```
optimal -> minimum

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/VertexCovers.java:
```diff
@@ -147,6 +147,60 @@
+     * @return a set of vertices which is an optimal vertex cover for the specified
+     * graph.
+     */
+    public static <V, E> Set<V> findVertexCover(Graph<V, E> g)
```
findVertexCover -> findMinimumVertexCover
To emphasize that you are not just looking for a general vertex cover, but a minimum one.

nolberg commented:
Thanks for your feedback. I took care of most of your comments, however, I could not come up yet with an efficient way that avoids constructing the two subgraphs. It might make sense to choose the vertex with highest degree instead of a random vertex, but I am not sure whether that improves the running time.

jkinable commented:
Thanks for your revision. Code review has revealed some issues that need to be addressed:

1. For your memoization, you cast the graph to a String using toString() and use that as a key.
  -This is incredibly expensive. In general you would never want to use toString().
  -It may result into (serious) errors! You assume that toString() creates a *unique* representation of the graph. This is certainly not a guarantee that toSTring() will give!

2. Unfixed from previous revision: you recurse on a graph, which is very expensive (see comments below).

3. There are no unit tests for this method. Simply testing whether the solution is a cover is not sufficient: you need to guarantee that it is indeed a minimum cover! Recommended tests:
a. test algorithm on some specific graph structures for which it is easy to compute (and verify) a minimum vertex cover: clique, star, odd cycle, even cycle.
b. create a number of random graphs of reasonable size e.g. 100 vertices or so. Use an external tool to calculate a minimum vertex cover for each of these graphs, thereby using e.g. an alternative implementation, or a Mixed Integer Programming model. Create a test case with the random graph as input, and compare against the hard-coded solution from the external tool.

4. Use the findGreedyCover() method to obtain a fast greedy solution. This solution can then be used to initialize your exact search, and to limit the depth of your recursion. 

5. Experiment with different vertex orderings. E.g. sort the vertices based on their degree in decending order. Calculate the vertex cover and measure the time it took. Repeat this for 10k random graphs, and report the average and maximum time. Then try some other sorting, e.g. vertex degree in ascending order. You can then compare Random ordering versus Vertex Degree Ascending ordering versus Vertex Degree Descending Ordering, etc.


I'm sorry to be so picky, but since JGraphT is a graph library, a lot of people depend on correct and efficient implementations in their own work. Furthermore, once an algorithm is added to the library, there is usually very little incentive for people to submit a more efficient version, as they assume that the submitted version is already 'good enough' or 'hard to improve'. As such, it is important to have reasonable high standards when it comes to accepting new submissions.


Here are some idea's to solve issues (1) and (2). I have no idea which approach to use without doing some further research and experimentation, but here are some suggestions.

1. You can create an ArrayList of the vertexSet thereby obtaining an ordered sequence of vertices. Next you can represent a "selection" of vertices as a java BitSet, where the number of bits equals the number of vertices. Each bit corresponds with a single vertex: bit[i]=0 means that vertex i is not selected, whereas bit[i]=1 means that vertex i is selected. This offers a compact relatively efficient and unique representation of a selection of vertices. Furthermore, computations can be very efficient since logic operations can be applied to bitsets.

2. Finding a minimum vertex cover has relatively strong similarities to a binary knapsack problem, where the problem is to select (or not) a number of items thereby maximizing the total value of the selected items, subject to a maximum weight constraint. See e.g.: http://www.personal.kent.edu/~rmuhamma/Algorithms/MyAlgorithms/Dynamic/knapsackdyn.htm  To speed up the Dynamic program, a c[i,w] table is used, where i is the item to select, and w is the remaining weight. Not sure whether this will work, but it's an interesting idea.

Nevertheless, I guess the easiest solution would be to look into some books/online for an existing algorithm to solve the min vertex cover problem and replicate that algorithm. That may be far less effort than implementing something from scratch.



Final remark: currently this class is implemented to find a minimum vertex cover in an undirected graph. Although this is a useful problem, a much more common problem is to find a vertex cover of minimum *weight*. In this problem, a positive weight is associated with each vertex. The goal is to find a vertex cover such that the total weight of the vertices in the cover is minimized. The problem you are currently solving is a 'special' case of this problem where each vertex has weight 1. In the current submission it is fine to focus on the special case, but try to make it as easy as possible for someone else to extend your code to deal with the more general weighted version of this problem.

nolberg commented:
I took care of 1. and 2., and I will work on the remaining issues later. I agree that casting a graph to a string might cause some serious errors. However, I now chose to cast a BitSet (that indicates which nodes are in the subgraph) to a string to identify the subgraph. This way it can be checked whether this subproblem was already solved. I thought that this might not be as dangerous (and expensive) as casting a graph to a string. Let me know if you think this is still an issue.

nolberg commented:
I added unit tests for random star, clique, cycle and bipartite graphs. Also, it turns out that sorting the vertices in descending order makes the algorithm perform roughly 1.5 times faster (which makes sense since the graph size decreases faster). 

Regarding 4): I am not sure how to add this. I believe it is necessary to try out all possibilities and not start from a given approximate solution, but I can be wrong. Just in case, here is the pseudocode I used originally to implement the algorithm.

```
VC(G):
if V = ∅ then return ∅
Choose an arbitrary node v ∈ G
G1 := (V − {v }, { e ∈ E | v ∈/ e })
G2 := (V − {v } − N (v ), { e ∈ E | e ∩ N (v ) = ∅ }) if |{v} ∪ VC(G1)| ≤ |N(v) ∪ VC(G2)|
then return {v} ∪ VC(G1)
else return N(v) ∪ VC(G2)
```

jkinable commented:
I went over the code. Unfortunately, many of the issues identified in the previous reviews still persist. A code review takes surprisingly much time, and this is getting to the point where doing the code reviews takes more time than just implementing the algorithm :(. Nevertheless, I do appreciate the time and effort you put into this, so below I'll try to give you some pointers to improve your future code, including some literature you could consult.

-you still use toString() as a unique key, which is still insecure and very slow. Reading: hashing and hash functions. Also read about String modifications. Strings in Java are immutable, so a toString() on a Collection object translates to a whole bunch of string concatenations (see http://stackoverflow.com/questions/10078912/best-practices-performance-mixing-stringbuilder-append-with-string-concat) which again is expensive and unnecessary.
-the algorithm should benefit from a quick greedy initial solution: this can limit the depth of the recursion.
-there seems to be a bug in your code: for some random graphs, I got vertex covers which were larger than the minimum vertex cover.
-there's a lot of overhead in your recursion. Consider what happens for example if you choose a vertex without any neighbors?
-You implemented some sort of inefficient sorting function to sort a list of vertices that runs in O(n^2) time. Reading: sorting algorithms in computer science. Aside from that, you shouldn't implement a sorting mechanism that runs in O(n^2) time; use the build-in sort mechanisms in java. Reading: Comparators in Java, Collections.sort and anonymous functions, e.g.:
	List<V> vertices=new ArrayList<>(g.vertexSet());
        Collections.sort(vertices, new VertexDegreeComparator<>(g, false));
	//OR USE:
        Collections.sort(vertices, (v1, v2) -> Integer.compare(g.degreeOf(v1), g.degreeOf(v2)));
-The code maintains a list of neighbors: use jgrapht's NeighborIndex for that instead
-While iterating over a bitset, you should not check every bit; instead, use the nextSetBit method (see javadoc for this method)
-The code doesn't come with proper test cases. Create some random graphs with 100-150 vertices, solve them using an external implementation of a vertex cover algorithm (e.g. Integer programming) and hardcode the graphs and corresponding solutions in a testcase. See MaximumFlowTestBase for an example.


In summary, I cannot afford spending more time on additional review iterations. However, since having an exact vertex cover algorithm is actually useful, I quickly implemented an algorithm which covers several of the above idea's: https://github.com/jkinable/jgrapht/blob/vertexCoverExact/jgrapht-core/src/main/java/org/jgrapht/alg/ExactVertexCover.java
 I would suggest that you have a look at the implementation, and that you use that implementation instead (just pull the file to your private repository). It scales quite well to graphs containing 150-250 vertices.
There are several things that have to be done before the pull request can be merged.
1. Adding proper test cases as indicated above. In addition, some special test cases on graphs with a special structure can be found in the main() function of my implementation (which needs to be deleted) 
2. Testing the implementation (may still contain bugs, didn't have time to do much testing).

Optional (I would recommend postponing these to a future pull request):
-you could extend the code with a new function Set<V> getMinimumVertexCover(Map<V,Integer> vertexWeights) which finds a minimum weight vertex cover, where the weight of the cover is the sum of the weights of the vertices, as provided by the weight map. This should be a trivial modification of the code.
-there are several 'TODOs' mentioned in the code. Some of them are however quite hard.
-a recursive implementation is rather simplistic and generic. I would be a good idea to look into some academic literature for dedicated algorithms. This may be a good start (I haven't read it myself): http://www.dharwadker.org/vertex_cover/
-It might be worth trying to replace the neighbors index by a bitset view. As such, all operations can be simplified to bitset operations, which may improve the algorithm's speed. The improvement may however be very marginal, and it will reduce the readability of the code.

nolberg commented:
@jkinable Thanks for taking the time to review my code and for suggesting an improved version. I guess since you created a pull request #239 for exact minimum vertex covers this one can be closed?

jkinable commented:
Pull request #239 does not include the exact algorithm yet. A minimal set of things that have to be done:
1. Refactor https://github.com/jkinable/jgrapht/blob/vertexCoverExact/jgrapht-core/src/main/java/org/jgrapht/alg/ExactVertexCover.java to make it compatible with pull request #239.
2. Add test cases:
- Special test cases which are commented out in the main function
- Tests with 100-150 vertex graphs, randomly generated and for which the solutions are calculated through some external means.

Since this is not much work, I'm happy to do these things myself in pull request #239, or leave you the honor of finishing this job. Let me know what you prefer. Either way, I'll add your name to the list of contributors.

jkinable commented:
Here's an idea. I'll merge the exact algorithm into  #239. That takes care of (1). Next, if you want, you can add the necessary tests and possible updates/changes to the exact algorithm, thereby taking care of (2). Easiest to do this is to submit a pull request into https://github.com/jkinable/jgrapht/tree/vertexCoverExact This immediately extends pull request #239. Would that work for you?

nolberg commented:
I am happy to help with (2). Just to be sure: were you thinking of randomly creating a few instances, computing an optimal solution for minimum vertex cover (e.g. using an LP solver), and then **hard code** those instances (with 100-150 vertices) in a test?

jkinable commented:
Yep that's it.
Maybe just to keep it as simple as:
1. Create 10 instances using the random graph generator. It's fine to make
10 instances, each of 100 vertices, with various numbers of edges
2. For each instance, create a vertex weight vector, containing a positive,
integer weight per vertex.
3. Use a Mixed Integer Programming (MIP) solver (or some other external
algorithm) to solve both an unweighted and a weighted version of the
problem. For the weighted version, use the weight vector created in step
(2), for the unweighted version, you can use a weight vector where all
vertices have weight 1.

4. Add the 10 weighted test graphs to WeightedVertexCoverTestjava, add the
10 unweighted test graphs to VertexCoverTest.java, under '// ------- Exact
algorithms ------'.

Have a look at the MaximumFlowAlgorithmTestBase.java class. Here you'll
find some examples of hardcoded graphs. A compact encoding would be:
int[][] edges={{1,2}, {2,3}, {8,5}, .....{5,20}}; //Hardcoded array with
all edges, size |E|
int[] weights={5, 3, 2, 1, ....}; //Hardcoded weight vector with an entry
for every vertex, size |V|=100
To build the graph:
SimpleWeightedGraph<Integer, DefaultEdge> graph=new
SimpleWeightedGraph<>(DefaultEdge.class);
for(int[] edge : edges) //Format: {u,v}
  Graphs.addEdgeWithVertices(graph, edge[0], edge[1]);
//Solve the instance
MinimumVertexCoverAlgorithm<Integer, DefaultEdge> mvc1=new
RecursiveExactVCImpl<>();
VertexCover<Integer> vertexCover=mvc1.getVertexCover(graph);
assertTrue(isCover(graph, vertexCover)); //Check whether the graph is a
vertex cover
assertEquals(vertexCover.getWeight(), <MIP Objective>); //Verify whether
the weight you computed using the external solver is identical to the
weight of the cover.


On Mon, Aug 1, 2016 at 5:19 PM, Nils Olberg <notifications@github.com>
wrote:

> I am happy to help with (2). Just to be sure: were you thinking of
> randomly creating a few instances, computing an optimal solution for
> minimum vertex cover (e.g. using an LP solver), and then *hard code*
> those instances (with 100-150 vertices) in a test?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jgrapht/jgrapht/pull/227#issuecomment-236711204>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACnYyjRp1ZhSRUncFwLfja7lbGKq8sTkks5qbmLugaJpZM4JTmbG>
> .
>

nolberg commented:
I have written [this](https://github.com/nolberg/scripts/blob/master/jgrapht/vertex_cover_lp_gen.py) Python script that does the following:
1. It creates a text file that contains java code, which is a test case for minimum weighted vertex cover.
2. It creates an .lp file that contains a linear program. This linear program models the minimum weighted vertex cover problem for the same instance as in the java code test case. 

The .lp file can be used to obtain an optimal solution using a solver. 

I have been adding a few instances to the other JGraphT tests, but it seems that the exact algorithm does not compute the correct solution (or there is something wrong with my script, although I could not find anything yet).

Sorry, I just recently started using GitHub and I am still not completely sure what is the right way to extend someone else's code. I submitted a pull request into https://github.com/jkinable/jgrapht/tree/vertexCoverExact, but so far I have not figured out how to contribute to your current version of JGraphT.

jkinable commented:
Hm, I'm not seeing any open pull requests. Don't worry about it, git can be
a bit daunting. Since the new vertex cover code is now in the master branch
of the jgrapht repository on github, you can as well open a new pull
request on the main branch as you did before. That's perhaps easiest to do.
Just to be sure:
-you are solving an Integer Program right, and not just a Linear Program;
the Vertex cover problem is not unimodular and as such cannot be solved
through an LP solver.
-you can always try hardcoding the solution you get out of your integer
program solver:
VertexCover<Integer> vc=new VertexCoverImpl(new
HashSet<>(Arrays.asList(1,2,4,5,6,7)), 123);
For this vertexCover you can then run the isCover(vc) method in the test
class.
-also, make sure you do not include that script in the jgrapht repository.
just add 10 hardcoded tests plus results to the test class as in the
example I gave.

On Wed, Aug 3, 2016 at 4:16 PM, Nils Olberg <notifications@github.com>
wrote:

> I have written this
> <https://github.com/nolberg/scripts/blob/master/jgrapht/vertex_cover_lp_gen.py>
> Python script that does the following:
> 1. It creates a text file that contains java code, which is a test case
> for minimum weighted vertex cover.
> 2. It creates an .lp file that contains a linear program. This linear
> program models the minimum weighted vertex cover problem for the same
> instance as in the java code test case.
>
> The .lp file can be used to obtain an optimal solution using a solver.
>
> I have been adding a few instances to the other JGraphT tests, but it
> seems that the exact algorithm does not compute the correct solution (or
> there is something wrong with my script, although I could not find anything
> yet).
>
> Sorry, I just recently started using GitHub and I am still not completely
> sure what is the right way to extend someone else's code. I submitted a
> pull request into
> https://github.com/jkinable/jgrapht/tree/vertexCoverExact, but so far I
> have not figured out how to contribute to your current version of JGraphT.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jgrapht/jgrapht/pull/227#issuecomment-237358609>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACnYyqUN4P0_vdv31SgPQ5hSwGJu-kJsks5qcPcggaJpZM4JTmbG>
> .
>

nolberg commented:
Yes, I am hoping that I will manage to solve the problems I am having with git right now.

And yes, I am solving an integer program (in fact there are only binary variables). I will have a look at it again later and try to figure out what the issue is.

nolberg commented:
There seems to be a problem with the getWeight() function of a vertex cover. For example, consider the following test: 


   
    public static void testExactMinimumCover1()
    {
	    int[] weightArray = {10, 7, 18, 4, 15, 9, 10, 14};
	
	    int[][] edges = {{0,6}, {1,5}, {1,6}, {1,7}, {3,7}};
	
	    UndirectedGraph<Integer, DefaultEdge> graph = new SimpleGraph<Integer, DefaultEdge>(DefaultEdge.class);
	
	    for (int[] edge : edges)
	    	Graphs.addEdgeWithVertices(graph, edge[0], edge[1]);
	
	    HashMap<Integer, Double> weights = new HashMap<Integer, Double>();
	    for (int i = 0; i < weightArray.length; i++)
	    	weights.put(i, Double.valueOf(weightArray[i]));
	
	    MinimumWeightedVertexCoverAlgorithm<Integer, DefaultEdge> mvc = new RecursiveExactVCImpl<Integer, DefaultEdge>();
	    VertexCover<Integer> vertexCover = mvc.getVertexCover(graph, weights);
	
	    assertTrue(vertexCover.getWeight() == 21);
     }

I computed the optimal value of a minimum weight vertex cover using a solver. An optimal cover is {0,1,3} with weight 21.

When I call ` System.out.println(vertexCover.getWeight());` before the assertTrue(), it prints `3.0`, although `System.out.println(vertexCover.getVertices().toString());` prints actually `[6, 7, 5]`. And the total weight of this vertex cover would in fact be 33 (sum of the last three vertices = 9 + 10 + 14 = 33).

jkinable commented:
Hm oops, I put a '==' instead of '!=' to identify weighted graphs. Should
be fixed now. Thanks!

On Wed, Aug 3, 2016 at 6:06 PM, Nils Olberg <notifications@github.com>
wrote:

> There seems to be a problem with the getWeight() function of a vertex
> cover. For example, consider the following test:
>
> public static void testExactMinimumCover1()
> {
>     int[] weightArray = {10, 7, 18, 4, 15, 9, 10, 14};
>
>     int[][] edges = {{0,6}, {1,5}, {1,6}, {1,7}, {3,7}};
>
>     UndirectedGraph<Integer, DefaultEdge> graph = new SimpleGraph<Integer, DefaultEdge>(DefaultEdge.class);
>
>     for (int[] edge : edges)
>         Graphs.addEdgeWithVertices(graph, edge[0], edge[1]);
>
>     HashMap<Integer, Double> weights = new HashMap<Integer, Double>();
>     for (int i = 0; i < weightArray.length; i++)
>         weights.put(i, Double.valueOf(weightArray[i]));
>
>     MinimumWeightedVertexCoverAlgorithm<Integer, DefaultEdge> mvc = new RecursiveExactVCImpl<Integer, DefaultEdge>();
>     VertexCover<Integer> vertexCover = mvc.getVertexCover(graph, weights);
>
>     assertTrue(vertexCover.getWeight() == 21);
>  }
>
> I computed the optimal value of a minimum weight vertex cover using a
> solver. An optimal cover is {0,1,3} with weight 21.
>
> When I call System.out.println(vertexCover.getWeight()); before the
> assertTrue(), it prints 3.0, although
> System.out.println(vertexCover.getVertices().toString()); prints actually [6,
> 7, 5]. And the total weight of this vertex cover would in fact be 33 (sum
> of the last three vertices = 9 + 10 + 14 = 33).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jgrapht/jgrapht/pull/227#issuecomment-237389957>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACnYyoqyHb1JutK4Goa49MtXO45v4qxRks5qcRDrgaJpZM4JTmbG>
> .
>

nolberg commented:
I added the tests for weighted and unweighted vertex covers, and your implementation passed all of them. I just changed the number of vertices from 100 to 60 . Otherwise, the algorithm took too long (at some point I forced it to stop). 

Should I close this pull request and submit a new one for the tests? Or should we just use this one?

jkinable commented:
Tests look good. Make sure the pull request only contains the files
VertexCoverTest.java and WeightedVertexCoverTest.java (no other files).
Also, replace assertTrue(vertexCover.getWeight() == 16); by
assertEquals(vertexCover.getWeight(),16.0);

Either modify this pull request, or close this one and submit a new clean
one.

On Wed, Aug 3, 2016 at 11:07 PM, Nils Olberg <notifications@github.com>
wrote:

> I added the tests for weighted and unweighted vertex covers, and your
> implementation passed all of them. I just changed the number of vertices
> from 100 to 60 . Otherwise, the algorithm took too long (at some point I
> forced it to stop).
>
> Should I close this pull request and submit a new one for the tests? Or
> should we just use this one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jgrapht/jgrapht/pull/227#issuecomment-237440399>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACnYytdAJ9nKQZYLeA1z6AoKo6-IbJCwks5qcVeHgaJpZM4JTmbG>
> .
>

nolberg commented:
I opened pull request #243 for the minimum (weight) vertex cover tests.
